src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,488 @@
+        {
+            if (username == null && (password != null || sessionToken != null))
+            {
+                throw new ArgumentException("A MONGODB-AWS must have access key id.");
+            }
+            if (username != null && password == null)
+            {
+                throw new ArgumentException("A MONGODB-AWS must have secret access key.");
+            }
+        }
+
+        // nested classes
+        private class AwsCredentials
+        {
+            public UsernamePasswordCredential Credentials;
+            public string SessionToken;
+        }
+
+        private class MongoAWSMechanism : ISaslMechanism
+        {
+            private readonly UsernamePasswordCredential _credential;
+            private readonly string _name;
+            private readonly IRandomByteGenerator _randomByteGenerator;
+            private readonly string _sessionToken;
+
+            public MongoAWSMechanism(
+                UsernamePasswordCredential credential,
+                string sessionToken,
+                IRandomByteGenerator randomByteGenerator)
+            {
+                _name = "MONGODB-AWS";
+                _credential = Ensure.IsNotNull(credential, nameof(credential));
+                _sessionToken = sessionToken;
+                _randomByteGenerator = Ensure.IsNotNull(randomByteGenerator, nameof(randomByteGenerator));
+            }
+
+            public string Name
+            {
+                get { return _name; }
+            }
+
+            public ISaslStep Initialize(IConnection connection, SaslConversation conversation, ConnectionDescription description)
+            {
+                Ensure.IsNotNull(connection, nameof(connection));
+                Ensure.IsNotNull(description, nameof(description));
+
+                var nonce = GenerateRandomBytes();
+
+                var document = new BsonDocument()
+                    .Add("r", new BsonBinaryData(nonce))
+                    .Add("p", new BsonInt32('n'));
+
+                var clientFirstMessageBytes = ToBytes(document);
+
+                return new ClientFirst(clientFirstMessageBytes, nonce, _credential, _sessionToken);
+            }
+
+            private byte[] GenerateRandomBytes()
+            {
+                return _randomByteGenerator.Generate(randomLength);
+            }
+        }
+
+        private static BsonDocument ToDocument(byte[] bytes)
+        {
+            MemoryStream stream = new MemoryStream(bytes);
+            using (var jsonReader = new BsonBinaryReader(stream))
+            {
+                var context = BsonDeserializationContext.CreateRoot(jsonReader);
+                return BsonDocumentSerializer.Instance.Deserialize(context);
+            }
+        }
+
+        private static byte[] ToBytes(BsonDocument doc)
+        {
+            BsonBinaryWriterSettings settings = new BsonBinaryWriterSettings()
+            {
+#pragma warning disable 618
```

Comment:
I guess you need to change this code on the following:

           if (BsonDefaults.GuidRepresentationMode == GuidRepresentationMode.V2)
            {
                clientSettings.GuidRepresentation = GuidRepresentation.Unspecified;
            }

Reply:
Done. Please recheck that it's what you expected.

Reply:
It's better to leave the suggested code since the reason why it's left without covering via methods or some other possible refactoring, that it will be removed in the version 3.0